internal/provider/resource_gitlab_pages_domain.go

```diff
@@ -0,0 +1,220 @@
+		UpdateContext: resourceGitlabPagesDomainUpdate,
+		DeleteContext: resourceGitlabPagesDomainDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"domain": {
+				Description: "The custom domain.",
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+			},
+			"project": {
+				Description: "The ID or full path of the project which the branch is created against.",
+				Type:        schema.TypeString,
+				ForceNew:    true,
+				Required:    true,
+			},
+			"auto_ssl_enabled": {
+				Description: "Enables automatic generation of SSL certificates issued by Let’s Encrypt for custom domains.",
+				Type:        schema.TypeBool,
+				Optional:    true,
+				ForceNew:    false,
+				Default:     nil,
+			},
+			"certificate": {
+				Description: "The certificate in PEM format with intermediates following in most specific to least specific order.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    false,
+				Default:     nil,
+			},
+			"key": {
+				Description: "The certificate key in PEM format.",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    false,
+				Default:     nil,
+			},
+			"url": {
+				Description: "The URL for the given domain.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+			"certificate_data": {
+				Description: "The certificate data.",
+				Type:        schema.TypeList,
+				MaxItems:    1,
+				Computed:    true,
+				Optional:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"expired": {
+							Description: "Is the certificate expired?",
+							Type:        schema.TypeBool,
+							Computed:    true,
+							Optional:    true,
+						},
+						"expiration": {
+							Description:  "The certificate expiration date.",
+							Type:         schema.TypeString,
+							ValidateFunc: validation.IsRFC3339Time,
+							Computed:     true,
+							Optional:     true,
+						},
+					},
+				},
+			},
+			"verified": {
+				Description: "The certificate data.",
+				Type:        schema.TypeBool,
+				Computed:    true,
+			},
+			"verification_code": {
+				Description: "The verification code for the domain.",
+				Type:        schema.TypeString,
+				Computed:    true,
+				Sensitive:   true,
+			},
+		},
+	}
+})
+
+func resourceGitlabPagesDomainCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*gitlab.Client)
+	projectId := d.Get("project").(string)
+	domain := d.Get("domain").(string)
+	auto_ssl_enabled := d.Get("auto_ssl_enabled").(bool)
+	certificate := d.Get("certificate").(string)
+	key := d.Get("key").(string)
+
+	options := &gitlab.CreatePagesDomainOptions{
+		Domain:         &domain,
+		AutoSslEnabled: &auto_ssl_enabled,
+		Certificate:    &certificate,
+		Key:            &key,
+	}
```

Comment:
NIT: to indicate that we only use these attributes for the GitLab options I think it makes sense to not have variables for them ...

```suggestion
	domain := d.Get("domain").(string)
	auto_ssl_enabled := d.Get("auto_ssl_enabled").(bool)
	certificate := d.Get("certificate").(string)
	key := d.Get("key").(string)

	options := &gitlab.CreatePagesDomainOptions{
		Domain:         gitlab.String(d.Get("domain").(string)),
		AutoSslEnabled: gitlab.Bool(d.Get("auto_ssl_enabled").(bool))
		Certificate:    gitlab.String(d.Get("certificate").(string)),
		Key:            gitlab.String(d.Get("key").(string)),
	}
```

Please verify the correct behavior during creation if these optional attributes are not given and that the GitLab API doesn't receive them ....

Reply:
Do you mean separate testcases? The current test does not specify any of the optional attributes.

Reply:
True, noticed afterwards - let's make sure that both scenarios are tested: creation without optionals set, creation with optionals set ...